### PR TITLE
[RFC] Expose fuse-svn from verifier logic and represent in DICE evidence.

### PR DESCRIPTION
**Pull request by bluegate010**
This commit also reorganizes the evidence across multiple DiceTcbInfos, to avoid overloading the `svn` field.

Partially addresses https://github.com/chipsalliance/caliptra-sw/issues/162

### CODE REVIEW

File: image/verify/src/verifier.rs
Comment (varuns-nvidia) [inline image/verify/src/verifier.rs:617]: Is this the same logic that BootROM will do for image authentication?

Comment (bluegate010) [reply image/verify/src/verifier.rs:617]: I would expect that boot ROM would not have to do the `MAX(min-SVN, SVN-fuses)` bit when authenticating the image.

This is where some feedback would be helpful. I went with MAX here because without it, if we just sample SVN-fuses, then on the first boot after min-SVN revs those fuses the DICE evidence may reflect what the fuses _were_, not what they are now (this assuming that SVN-fuses get blown after Caliptra boots, not before). A subsequent power cycle would cause the evidence to catch up to the fuse state.

Comment (varuns-nvidia) [reply image/verify/src/verifier.rs:617]: There's multiple layers here because what Caliptra sees are registers and not the fuses themselves. The registers are latched on Caliptra power-good and locked as part of the Boot FSM transition states. So there's no way for Caliptra ROM to know any fuse state change until the next power-good anyway.

Either ROM or FW acting as update authorizer needs to do the `MAX(min-SVN, SVN-fuses)` logic to handle hitless update cases, because min-SVN can rev on update reset via manifest, but SVN-fuses cannot. I prefer ROM does it so we're not implementing authorization logic in different places.

Comment (bluegate010) [reply image/verify/src/verifier.rs:617]: I see; it sounds like in either case, we should be writing this effective-min-SVN out to the DataVault.

Should we try and create a mechanism for Caliptra ROM to signal to SoC what the new value of SVN-fuses should be?

Also, I gather there's an assumption here that upon hitless update, while we can increment some "soft" min-SVN and stash it in the DataVault, we can't increment the "hard" min-SVN stored in fuses?

Comment (varuns-nvidia) [reply image/verify/src/verifier.rs:617]: > Should we try and create a mechanism for Caliptra ROM to signal to SoC what the new value of SVN-fuses should be?

That can be useful, but it's also possible for the SoC to just manage the fuses orthogonally from what Caliptra says.

> Also, I gather there's an assumption here that upon hitless update, while we can increment some "soft" min-SVN and stash it in the DataVault, we can't increment the "hard" min-SVN stored in fuses?

Yep. @vsonims @bharatpillilli for vis

Comment (bluegate010) [reply image/verify/src/verifier.rs:617]: Ack. So we should think about what kind of min-SVN we actually want to report. If we report min-SVN as present in the firmware header, then that's a directive to firmware saying "fuses should be at least this", but the fuses might not have been blown by the time Caliptra runs, and Caliptra can't see what its new value is, as it doesn't manage fuses. Also if fuses are X and min-SVN is X-1, that's no less trustworthy a state than if min-SVN were X. So ignoring fuses seems a bit weird.

We could report what the fuse values were at boot-up. If min-SVN caused fuses to increment in the current boot cycle, a power cycle would be needed to actually get the attested evidence to update.

We could report the soft-SVN latched in the data vault. We _can't_ report that on FMC's alias key cert, though, because we don't re-generate that across hitless updates. Would need FMC to sample it and put it in RT FW's alias cert in order for it to reflect reality.

I'm not sure what all would be interesting to a verifier. @ericeilertson, any thoughts?

Comment (bluegate010) [reply image/verify/src/verifier.rs]: Resolving this; we now use fuse-SVN as per https://github.com/chipsalliance/caliptra-sw/issues/269

**Review by FerralCoder — APPROVED** [review]
I think this looks good but I would like @fdamato to confirm some things that we had discussed earlier today before this gets merged. He said he will be able to do that sometime today.

File: image/verify/src/verifier.rs
Comment (fdamato) [inline image/verify/src/verifier.rs:608]: I`m not sure I understand how this works. Per spec Runtime SVN is 128 bits, and FMC is 32 bits.
Also it gets later downcast to a u8

Comment (bluegate010) [reply image/verify/src/verifier.rs:608]: The bits start out in a unary representation, but the driver [converts](https://github.com/chipsalliance/caliptra-sw/blob/2e26d715dcbfa92c1e5e7eb3bcd2cda760db9792/drivers/src/fuse_bank.rs#L206-L236) them to numerical.

I picked u32 as the type just to mirror what's already present at this layer, but we could just change this to u8 here, to cut down on the number of different representations.

Comment (fdamato) [reply image/verify/src/verifier.rs:608]: I think sticking with either u32 or u8 might prevent any room for confusion...but I`m ok with that. Thanks

**Review by fdamato — APPROVED** [review]

File: image/verify/src/verifier.rs
Comment (bluegate010) [reply image/verify/src/verifier.rs:608]: Ack, will refactor to use u8 consistently in a follow-up PR (don't have bandwidth this week and I understand there's work that's blocked on this).

**Pull request merged**
